tools/golden_tests_harvester/lib/golden_tests_harvester.dart

```diff
@@ -5,77 +5,176 @@
+    final io.File file = io.File(p.join(workDirectory.path, 'digest.json'));
+    if (!file.existsSync()) {
+      // Check if the directory exists or if the file is just missing.
+      if (!workDirectory.existsSync()) {
+        throw ArgumentError('Directory not found: ${workDirectory.path}.');
+      }
+      // Lookup sibling files to help the user understand what's missing.
+      final List<io.FileSystemEntity> files = workDirectory.listSync();
+      throw StateError(
+        'File "digest.json" not found in ${workDirectory.path}.\n\n'
+        'Found files: ${files.map((io.FileSystemEntity e) => p.basename(e.path)).join(', ')}',
+      );
+    }
+    final Digests digests = Digests.parse(file.readAsStringSync());
+
+    if (addImageToSkiaGold != null) {
+      return _VanillaHarvester(digests, stderr, workDirectory, addImageToSkiaGold);
+    } else {
+      return SkiaGoldHarvester._create(digests, stderr, workDirectory);
     }
-    // Lookup sibling files to help the user understand what's missing.
-    final List<io.FileSystemEntity> files = workDirectory.listSync();
-    throw StateError(
-      'File "digest.json" not found in ${workDirectory.path}.\n\n'
-      'Found files: ${files.map((io.FileSystemEntity e) => p.basename(e.path)).join(', ')}',
-    );
   }
-  final Digests digests = Digests.parse(file.readAsStringSync());
+
+  Future<void> _addImg(
+    String testName,
+    io.File goldenFile, {
+    double differentPixelsRate,
+    int pixelColorDelta,
+    required int screenshotSize,
+  });
+
+  Future<void> _auth();
+
+  Digests get _digests;
+  StringSink get _stderr;
+  io.Directory get _workDirectory;
+}
+
+/// A [Harvester] that communicates with a real [SkiaGoldClient].
+class SkiaGoldHarvester implements Harvester {
+  SkiaGoldHarvester._init(
+      this._digests, this._stderr, this._workDirectory, this.client);
+
+  @override
+  final Digests _digests;
+  @override
+  final StringSink _stderr;
+  @override
+  final io.Directory _workDirectory;
+  /// The [SkiaGoldClient] that will be used for harvesting.
+  final SkiaGoldClient client;
+
+  static Future<SkiaGoldHarvester> _create(
+      Digests digests, StringSink stderr, io.Directory workDirectory) async {
+    final SkiaGoldClient client =
+        SkiaGoldClient(workDirectory, dimensions: digests.dimensions);
+    return SkiaGoldHarvester._init(digests, stderr, workDirectory, client);
+  }
+
+  @override
+  Future<void> _addImg(String testName, io.File goldenFile,
+      {double differentPixelsRate = 0.01,
+      int pixelColorDelta = 0,
+      required int screenshotSize}) async {
+    return client.addImg(testName, goldenFile,
+        differentPixelsRate: differentPixelsRate,
+        pixelColorDelta: pixelColorDelta,
+        screenshotSize: screenshotSize);
+  }
+
+  @override
+  Future<void> _auth() {
+    return client.auth();
+  }
+}
+
+/// A [Harvester] that doesn't harvest, just logs to [stderr].
+class _VanillaHarvester implements Harvester {
```

Comment:
nit: Maybe just `_DryRunHarvester` (since that's it's purpose)

Reply:
done